### PR TITLE
Add typeof expression support

### LINKED ISSUE #22432: Add typeof expression
**Description:**

Refer: https://ballerina.io/spec/lang/2020R1/#typeof-expr

### PR DESCRIPTION
**Pull request by lochana-chathura**
Fixes #22432 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by lochana-chathura** [comment]
some checks are failing. Will fix

**Comment by lochana-chathura** [comment]
Azure failure is not from my side.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 1]
```diff
@@ -2034,6 +2037,8 @@ private SyntaxKind getExpectedTokenKind(ParserRuleContext ctx) {
                 return SyntaxKind.FINAL_KEYWORD;
             case CONST_DECL_TYPE:
                 return SyntaxKind.IDENTIFIER_TOKEN;
+            case TYPEOF_KEYWORD:
+                return SyntaxKind.TYPEOF_KEYWORD;
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:2041]: Need to add a case for `getNextRule` as well.

Try this one: `int a = 3  typeof  + 2;`.  Might fail.

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:2043]: Done

Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:2043]: shall we add a test-case too for this particular scenario?

**Comment by lochana-chathura** [comment]
one recovery test in record-type-def failing due to new changes. I will look into that

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java (hunk 1 above)
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:2075]: test case added for `int a = 3 typeof 2;`

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 2]
```diff
@@ -1787,6 +1789,8 @@ private OperatorPrecedence getOpPrecedence(SyntaxKind binaryOpKind) {
                 return OperatorPrecedence.LOGICAL_AND;
             case LOGICAL_OR_TOKEN:
                 return OperatorPrecedence.LOGICAL_OR;
+            case TYPEOF_KEYWORD:
+                return OperatorPrecedence.UNARY;
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:1793]: According to OperatorPrecedence.java `UNARY(1), //  (+x), (-x), (~x), (!x), (<T>x), (typeof x), (check x), (checkpanic x)`
Therefore, `check` and `checkpanic` cases should also have declared with `OperatorPrecedence.UNARY;` 
but in ballerina-spec document they are not listed in the precedence table.

**Pull request merged**